src/main/java/appeng/fluids/helper/DualityFluidInterface.java

```diff
@@ -0,0 +1,632 @@
+	}
+
+	@Override
+	public TickRateModulation tickingRequest( final IGridNode node, final int ticksSinceLastCall )
+	{
+		if( !this.gridProxy.isActive() )
+		{
+			return TickRateModulation.SLEEP;
+		}
+
+		final boolean couldDoWork = this.updateStorage();
+		return this.hasWorkToDo() ? ( couldDoWork ? TickRateModulation.URGENT : TickRateModulation.SLOWER ) : TickRateModulation.SLEEP;
+	}
+
+	public void notifyNeighbors()
+	{
+		if( this.gridProxy.isActive() )
+		{
+			try
+			{
+				this.gridProxy.getTick().wakeDevice( this.gridProxy.getNode() );
+			}
+			catch( final GridAccessException e )
+			{
+				// :P
+			}
+		}
+
+		final TileEntity te = this.iHost.getTileEntity();
+		if( te != null && te.getWorld() != null )
+		{
+			Platform.notifyBlocksOfNeighbors( te.getWorld(), te.getPos() );
+		}
+	}
+
+	public void markDirty()
+	{
+		// rescan inventory
+		for( IFluidHandler ih : this.tanks )
+		{
+			onFluidInventoryChanged( ih );
+		}
+	}
+
+	public void gridChanged()
+	{
+		try
+		{
+			this.items.setInternal( this.gridProxy.getStorage().getInventory( AEApi.instance().storage().getStorageChannel( IItemStorageChannel.class ) ) );
+			this.fluids.setInternal( this.gridProxy.getStorage().getInventory( AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ) ) );
+		}
+		catch( final GridAccessException gae )
+		{
+			this.items.setInternal( new NullInventory<IAEItemStack>() );
+			this.fluids.setInternal( new NullInventory<IAEFluidStack>() );
+		}
+
+		this.notifyNeighbors();
+	}
+
+	public AECableType getCableConnectionType( final AEPartLocation dir )
+	{
+		return AECableType.SMART;
+	}
+
+	public DimensionalCoord getLocation()
+	{
+		return new DimensionalCoord( this.iHost.getTileEntity() );
+	}
+
+	public boolean hasCapability( Capability<?> capabilityClass, EnumFacing facing )
+	{
+		return capabilityClass == CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY || capabilityClass == Capabilities.STORAGE_MONITORABLE_ACCESSOR;
+	}
+
+	@SuppressWarnings( "unchecked" )
+	public <T> T getCapability( Capability<T> capabilityClass, EnumFacing facing )
+	{
+		if( capabilityClass == CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY )
+		{
+			return (T) this.storage;
+		}
+		else if( capabilityClass == Capabilities.STORAGE_MONITORABLE_ACCESSOR )
+		{
+			return (T) this.accessor;
+		}
+		return null;
+	}
+
+	private boolean hasConfig()
+	{
+		return this.hasConfig;
+	}
+
+	private void readConfig()
+	{
+		final boolean had = this.hasWorkToDo();
+
+		for( int x = 0; x < NUMBER_OF_TANKS; x++ )
+		{
+			this.configStacks[x] = null;
+			ItemStack is = this.config.getStackInSlot( x );
+			if( !is.isEmpty() )
+			{
+				IFluidHandlerItem fh = is.getCapability( CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null );
+				if( fh != null )
+				{
+					FluidStack fs = fh.drain( Fluid.BUCKET_VOLUME, false );
+					if( fs != null && fs.amount > 0 )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)